apps/src/lib/tools/jsinterpreter/CustomMarshaler.js

```diff
@@ -0,0 +1,69 @@
+
+module.exports = class CustomMarshaler {
+  constructor({
+    globalProperties,
+    blockedProperties,
+    objectList,
+  }) {
+    this.globalProperties = globalProperties || {};
+    this.blockedProperties = blockedProperties || [];
+    this.objectList = objectList || [];
+  }
+
+  // If this is on our list of "custom marshal" objects - or if it a property
+  // on one of those objects (other than a function), return true
+  shouldCustomMarshalObject(nativeVar, nativeParentObj) {
+    for (var i = 0; i < this.objectList.length; i++) {
+      var marshalObj = this.objectList[i];
+      if ((nativeVar instanceof marshalObj.instance &&
+           (typeof marshalObj.requiredMethod === 'undefined' ||
+            nativeVar[marshalObj.requiredMethod] !== undefined)) ||
+          (typeof nativeVar !== 'function' &&
+           nativeParentObj instanceof marshalObj.instance)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // When marshaling methods on "custom marshal" objects, we may need to augment
+  // the marshaling options. This returns those options.
+  getCustomMarshalMethodOptions(nativeParentObj) {
+    for (var i = 0; i < this.objectList.length; i++) {
+      var marshalObj = this.objectList[i];
+      if (nativeParentObj instanceof marshalObj.instance) {
+        if (typeof marshalObj.requiredMethod === 'undefined' ||
+            nativeParentObj[marshalObj.requiredMethod] !== undefined) {
+          return marshalObj.methodOpts || {};
+        } else {
+          return {};
```

Comment:
Having just extracted this helper, it'd be nice for it to have 100% test coverage.